certora/spec/yieldBox.spec

```diff
@@ -75,369 +60,588 @@ methods {
-            require assetId == getAssetId(asset);
-            require assetsIdentical1(getAssetId(asset), asset);
-            require asset.tokenType == YieldData.TokenType.ERC721;
-            require dummyERC721 == asset.contractAddress;
-            require totalSupply(e2, getAssetId(asset)) <= 1;
+        preserved{
+            require getAssetsLength() < 1000000;
         }
     }
 
 
-invariant nftSharesEQzeroTest(YieldData.Asset asset, env e)
-    totalSupply(e,getAssetId(asset)) == 1 
-        => (dummyERC721.ownerOf(e,asset.tokenId) == YieldData ||
-            dummyERC721.ownerOf(e,asset.tokenId) == asset.strategy)
-    
-    filtered { f -> f.selector != batch(bytes[],bool).selector 
-                    && f.selector != uri(uint256).selector 
-                    && f.selector != name(uint256).selector 
-                    && f.selector != symbol(uint256).selector 
-                    && f.selector != decimals(uint256).selector  }
-    {
-        preserved with (env e1){
-            require assetsIdentical1(getAssetId(asset),asset);
-            require asset.tokenType == YieldData.TokenType.ERC721;
-            require dummyERC721 == asset.contractAddress;
-        }
-        preserved depositNFTAsset(uint256 assetId, address from, address to) with (env e2) {
-            require assetId == getAssetId(asset);
-            require assetsIdentical1(getAssetId(asset), asset);
-            require asset.tokenType == YieldData.TokenType.ERC721;
-            require dummyERC721 == asset.contractAddress;
-        }
-    }
+// STATUS - verified
+// shares solvency: sum of shares of all users should be less than or equal to totalSupply of shares.
+invariant sharesSolvency()
+    forall uint256 id. sharesSum[id] <= totalSupplyGhost[id];
 
 
 
-invariant tokenTypeValidity(YieldData.Asset asset, env e)
-    getAssetTokenType(getAssetId(asset)) == 4 => _tokenBalanceOf(e, asset) == 0
+////////////////////////////////////////////////////////////////////////////
+//                       Rules                                            //
+////////////////////////////////////////////////////////////////////////////
 
-// check for correctness
-invariant balanceVStotalSupply(YieldData.Asset asset, env e)
-     _tokenBalanceOf(e,asset) == 0 => totalSupply(e,getAssetId(asset)) == 0
 
-// in progress
+// STATUS - violation becuase withdrawNFT allows to pass 0 as shares and amount but still burns it 
+// you can pass 0's to withdraw if you withdraw NFT. but something will be burnt anyway becuase it's hardcoded
 // Integrity of withdraw()
-rule withdrawIntegrity() 
+rule withdrawIntegrity(env e) 
 {
-    env e;
-
     uint amountOut; uint shareOut;
     uint amount; uint share;
     address from; address to;
 
-    YieldData.Asset asset;
-    uint assetId = getAssetId(asset);
-
-    require asset.strategy == Strategy;
+    YieldBoxHarness.Asset asset;
+    uint assetId;
+    require assetsIdentical1(assetId, asset);
```

Comment:
yes, otherwise induction creates a mess that violates any property easily. Any reason why I don't need it? I know that you are not familiar with the code, but from your perspective?

Reply:
that maybe the require after that is not necessary after this one

Reply:
both of these requirements are needed because assets are stored in 
- mapping where asset's fields point to asset's id
- array where asset's id points to asset's fields
It's kind of a way to synch it. It was implemented long time ago and making it look pretty will take a lot of time.